compiler/vm/vmdef.nim

```diff
@@ -90,27 +91,224 @@ type
+    slotTempStr,      ## some temporary string
+    slotTempComplex,  ## some complex temporary (s.node field is used)
+    slotTempPerm      ## slot is temporary but permanent (hack)
+
+  AtomKind* = enum
+    # Trivial values
+    akInt
+    akFloat
+    akPtr
+
+    # Managed values
+    akSet
+    akString
+    akSeq
+    akRef
+    akCallable # XXX: rename to akProcedural or akFuncHandle
+
+    akDiscriminator
+
+    akPNode # XXX: This is two things, typedesc values in the context
+            # of type evaluation and `NimNode`
+            # Since these are somewhat different, and `typedesc` can't be stored in locations,
+            # it might make sense to introduce a new register kind for typedescs...
+
+    # Pseudo atom kinds
+    akObject
+    akArray
+
+
+  # XXX: `ref` until I decide on a better way to manage these
+  PVmType* = ref VmType
+
+  FieldPosition* = uint32
+
+  BranchListEntryKind* = enum
+    blekStart
+    blekBranch
+    blekEnd
+
+  BranchListEntry* = object
+    case kind*: BranchListEntryKind
+    of blekStart:
+      field*: FieldPosition
+      numItems*: uint32
+      numBranches*: uint16
+    of blekBranch:
+      fieldRange*: Slice[FieldPosition]
+    of blekEnd:
+      parent*: uint32
+      nextField*: FieldPosition
+
+  VmType* = object ## A `VmType` describes the type of a location
+    case kind*: AtomKind
+    of akInt, akFloat, akString, akPNode: discard
+    of akPtr, akRef:
+      targetType*: PVmType
+    of akSet:
+      discard
+    of akSeq:
+      seqElemStride*: int
+      seqElemType*: PVmType
+    of akCallable:
+      retType*: PVmType
+    of akDiscriminator:
+      # A discriminator consists of two things: the value as seen by the guest
+      # and the index of the branch
+      numBits*: int ## The amount of bits the discriminator value occupies
+    of akObject:
+      # TODO: offset could be turned into a uint32 (or even uint16).
+      # But due to alignment of `typ`, it probably wouldn't buy us anything 
+      objFields*: seq[tuple[offset: int, typ: PVmType]]
+
+      branches*: seq[BranchListEntry]
+    of akArray:
+      elementCount*: int
+      elementStride*: int ## The amount of bytes between the start of two consecutive 
+                         ## elements. Can be larger than the size of an element alone
+      elementType*: PVmType
+
+    sizeInBytes*: uint
+    alignment*: uint8 ## 1 shl `alignment` == real alignment value
+
+  CallableKind* = enum
+    ckDefault ## A referenc to a VM function
+    ckClosure
+    ckCallback ## A VmCallback
+    ckFFI
+    ckTemplate
+
+  # XXX: design is temporary
+  VmFunctionObject* = object ## An object storing a callable
+    case kind*: CallableKind
+    of ckDefault:
+      prcSym*: PNode # TODO: needs to be a `PSym`
+    of ckClosure:
+      closureEnv*: GlobalRef ## A reference to the closure environment
+      closureTarget*: int
+    of ckCallback:
+      cbOffset*: int
+    of ckFFI:
+      ffiPrcSym*: PSym
```

Comment:
What kind of FFI does this imply? C FFI via `libffi` or VM callbacks?

In the case of `libffi`, I think you can left it off the table for now, as the tests for it are experimental and not run at all.

Reply:
Yes. The previous VM had support for C FFI via `libffi`, which is why I also included it here. If I understood you correctly, I'm allowed to remove it for now?

Reply:
Yea, it's not even tested nor enabled by default, so I'd write it off as dead code.

Reply:
Seconded, it's a pretend feature. It was hacked in and it's not just dead weight it's actively harmful. If you spot other things like that please flag them.

Difficult algorithms, fundamental properties, broadly clean structure, correctness by design, debugging facilities, and a well specified tests are valuable. The rest can be reimplemented rather easily.

Sadly a good heuristic is if it's implemented by Tim be very skeptical about its quality.